Callout: Add component

norrise120 opened the pull request:
<!--
What is the purpose of this PR? The purpose of this PR is to add a new Callout component 

* What is the context surrounding this PR? Include links if possible.
* What kind of feedback do you want?
* Have you [formatted the PR title](https://github.com/pinterest/gestalt/#releasing)? `ComponentName: Description`
-->

## Test Plan
Wrote tests for component, matching pattern of testing for other components. 
Updated docs and verify that docs looked as expected.

<img width="840" alt="Screen Shot 2020-07-24 at 10 10 24 AM" src="https://user-images.githubusercontent.com/46180742/88417063-189dba80-cd96-11ea-852f-760dedf85883.png">


<!--
How can reviewers verify this is good to merge?

* Is it tested?
* Is it accessible?
* Is it documented?
* Have you involved other stakeholders (such as a Pinterest Designer)?
-->

norrise120 commented on docs/src/components/sidebarIndex.js:
```diff
@@ -2,14 +2,14 @@
-  displayName: string,
-  pages?: Array<string>,
-  subsections?: Array<{| [string]: Array<string> |}>,
+    displayName: string,
+      pages ?: Array < string >,
+      subsections ?: Array < {| [string]: Array<string> |}>,
```
My editor kept doing this when I saved. I'll update when I address feedback and make sure to remove these changes

norrise120 commented on packages/gestalt/src/Callout.js:
```diff
@@ -0,0 +1,111 @@
+import typography from './Typography.css';
+
+type Props = {|
+  description: string,
+    iconAccessibilityLabel: string,
+      primaryLink ?: {|
+        href: string,
+          label: string,
+  |},
+onDismiss: () => void,
+  secondaryLink ?: {|
+    href: string,
+      label: string,
+  |},
+style: 'error' | 'info' | 'warning',
+  title ?: string,
```
My editor did this whenever I saved. I'll update to make sure it's formatted correctly when I address feedback

christianvuerings commented on docs/src/Callout.doc.js:
```diff
@@ -0,0 +1,100 @@
+    name="Info Example"
+    defaultCode={`
+<Callout
+    style="info" 
```
@norrise120 could you enforce 2 spaces indentation? (just like you did in the example below)

christianvuerings commented on docs/src/Callout.doc.js:
```diff
@@ -0,0 +1,100 @@
+    defaultCode={`
+<Callout
+    style="info" 
+    iconAcessibilityLabel=""
```
@norrise120 let's fill this out, use something like `Dismiss banner`

christianvuerings commented on docs/src/Callout.doc.js:
```diff
@@ -0,0 +1,100 @@
+      },
+      {
+        name: 'primaryLink',
+        type: '{| href: string, label: string |}',
```
@norrise120 could we also add an `onClick` handler? That way we can integrate these links with react-router in the Pinterest code base.

Try to use the same type for `onClick` as we use on `<Link />`:

```
  onClick?: AbstractEventHandler<
    | SyntheticMouseEvent<HTMLAnchorElement>
    | SyntheticKeyboardEvent<HTMLAnchorElement>
```

christianvuerings commented on packages/gestalt/src/Callout.js:
```diff
@@ -0,0 +1,111 @@
+  info: {
+    icon: 'check-circle',
+    color: 'blue',
+    backgroundColor: '#EBF4FE',
```
@ashleyseto @anniemochi could we spec out what these colors should be in dark mode?

ashleyseto replied:
At the moment, there is not a dark mode spec for this as it is a non-standard color so I would leave it identical in dark mode to ship.

christianvuerings replied:
This is what currently happens in dark mode:
![image](https://user-images.githubusercontent.com/127199/88546811-c21cc000-cfd1-11ea-9028-535afb74e821.png)

So we do need to figure out a way forward. Right now the text colors and everything else automatically updates. The easiest way forward would be to have dark mode specific colors for these non-standard colors.

ashleyseto replied:
Looks like more than just color isn't right in that above screenshot. My immediate recommendation since dark mode was not in scope when this was designed would be that it would look identical in dark and light mode much like primary buttons are the same in both.

christianvuerings commented on packages/gestalt/src/Callout.test.js:
```diff
@@ -0,0 +1,84 @@
+        title="A Title"
+      />
+    ).toJSON();
+    expect(tree).toMatchSnapshot();
```
@norrise120 thanks for adding these tests!

christianvuerings commented on packages/gestalt/src/Callout.test.js:
```diff
@@ -0,0 +1,84 @@
+    ).toJSON();
+    expect(tree).toMatchSnapshot();
+  });
+});
```
@norrise120 could we also add a `.jsdom` test which tests the dismiss functionality?

ashleyseto commented:
Hey. Thanks for taking this on! A few things that need adjustment based on spec:

- The header (bold text) should be heading small
- Icon size should be 32px
- There is some padding around the icon button to close. Currently it is 16px in spec but we might need to adjust slightly on how it looks in code.
- 8px of space should be between the buttons for the example
- I'd like one of the examples to be without an x to close. We have the info, error, warning and the minimal option without an x.

ashleyseto commented:
Tagging in @ajmoo who did the original design spec for a final pass. 

One other adjustment, the text block is 24px from the icon and the full width of the text block is max 600px.

ajmoo commented:
aside from what @ashleyseto called out this looks good to me. "oops" is my favorite.

ashleyseto commented:
<img width="1447" alt="Screen Shot 2020-07-27 at 8 14 22 AM" src="https://user-images.githubusercontent.com/2187170/88559188-5f332500-cfe1-11ea-9670-113495f5fe92.png">

Hey a few adjustments.

- The 24px was meant to be in addition to the 600px specified for the text box width. Sorry that wasn't clear. The total space for text and padding of that box then would be 648px
- The buttons are still meant to be aligned to the right of the callout as the component grows in size.

ashleyseto commented:
<img width="411" alt="Screen Shot 2020-07-27 at 9 17 55 AM" src="https://user-images.githubusercontent.com/2187170/88566072-55fa8600-cfea-11ea-90be-2c33fb4b2a4e.png">

Chatted with Taylor and it seems we need to add a different look to for the gestalt-sm breakpoint for m-web. The buttons need to shift under the textbox and the icon would align to the top. Added to the Figma as well.

ashleyseto commented:
@taylorannsimpson for visibility

norrise120 commented:
Updated darkmode so that text is readable. I left the button untouched, but I can update to make it match the lightmode version exactly if preferred. 
<img width="1014" alt="Screen Shot 2020-07-27 at 11 07 49 PM" src="https://user-images.githubusercontent.com/46180742/88630148-09a05c00-d065-11ea-883c-1b6ae6082049.png">

> <img alt="Screen Shot 2020-07-27 at 9 17 55 AM" width="411" src="https://user-images.githubusercontent.com/2187170/88566072-55fa8600-cfea-11ea-90be-2c33fb4b2a4e.png">
> 
> Chatted with Taylor and it seems we need to add a different look to for the gestalt-sm breakpoint for m-web. The buttons need to shift under the textbox and the icon would align to the top. Added to the Figma as well.

If ok with everyone, I would prefer to handle the mweb updates in a different PR later this week/early next week.

christianvuerings commented:
> Updated darkmode so that text is readable. I left the button untouched, but I can update to make it match the lightmode version exactly if preferred.
> <img alt="Screen Shot 2020-07-27 at 11 07 49 PM" width="1014" src="https://user-images.githubusercontent.com/46180742/88630148-09a05c00-d065-11ea-883c-1b6ae6082049.png">
> 
That's totally fine

christianvuerings commented on packages/gestalt/src/Callout.js:
```diff
@@ -0,0 +1,159 @@
+      {secondaryLink && (
+        <Box paddingX={1} marginStart="auto">
+          <Box padding={4} flex="none">
+            <Link href={secondaryLink.href}>
```
@norrise120 could we add the `onClick` we passed through? `onClick={secondaryLink.onClick}`

ashleyseto commented:
One other minor change. The padding on the right where the buttons are should be 48px.

Current
<img width="1109" alt="Screen Shot 2020-07-28 at 11 08 00 AM" src="https://user-images.githubusercontent.com/2187170/88704225-a1806300-d0c2-11ea-9546-056188751ff5.png">

Desired
<img width="1114" alt="Screen Shot 2020-07-28 at 11 07 50 AM" src="https://user-images.githubusercontent.com/2187170/88704289-b826ba00-d0c2-11ea-86ce-5bf5db2b1edd.png">

We'll handle the responsive part of the component in a separate PR then. Thanks :)

AlbertCarreras commented:
Notes:

**API**
- is iconAccessibilityLabel needed? I've seen other components not adding accessibility labels to internal icons.
- should primaryLink focus the box instead of the link? It's a confusing linkButton.
- primaryLink/secondaryLin, should they allow target?

**Docs**
- We need a brief description
